speech/infinite_streaming_from_file/infinite_audio.go

```diff
@@ -0,0 +1,236 @@
+				}
+
+				//Append the new audio data to processingBuffer for future comparison
+				processingBuffer = append(processingBuffer, [][]byte{framesPerBuffer}...)
+
+				//Send new requests
+				if len(framesPerBuffer) > 0 {
+					if err := stream.Send(&speechpb.StreamingRecognizeRequest{
+						StreamingRequest: &speechpb.StreamingRecognizeRequest_AudioContent{
+							AudioContent: framesPerBuffer,
+						},
+					}); err != nil {
+						log.Printf("Could not send audio: %v", err)
+					}
+				}
+
+				//Control sending rate(send 1 second's audio data per second) and make sure it is the same as the actual audio play speed
+				time.Sleep(1 * time.Second)
+			}
+		}()
+
+		//Receive and process responses
+		exitFlag = func() bool {
+			for {
+				//Receive responses from server
+				resp, err := stream.Recv()
+
+				//Exception Handling:
+				//If server has sent all the responses ,then it will return EOF(End of File) error to the client
+				if err == io.EOF {
+					break
+				}
+
+				//If server couldn't send back responses, then it will return detailed error information
+				if err != nil {
+					log.Fatalf("Cannot stream results: %v", err)
+				}
+
+				//If server couldn't recognize audio data, then it will return detailed error information
+				if err := resp.Error; err != nil {
+					log.Fatalf("Could not recognize: %v", err)
+				}
+
+				//Process responses
+				for _, result := range resp.Results {
+					//If server has sent back final transcript for specific audio chunks, then calculate the end time pointer
+					//and remove audio chunks which have obtained final transcript already from processingBuffer
+					if result.GetIsFinal() {
+						transcript := result.Alternatives[0].Transcript
+						color.Blue("Transcript: %v\n", transcript)
+
+						//Calculate the correct end time of the specific response which contains final transcript based on restart times
+						resultEndTime := result.GetResultEndTime().Seconds*1000 + int64(math.Round(float64(result.GetResultEndTime().Nanos/1000000))) + int64(int(streamTimeLimit.Seconds())*1000*(restartCounter-1))
+
+						//Remove audio chunks which have obtained final transcript already
+						for len(processingBuffer) > 0 {
+							//Every element of processingBuffer is one sample which stores 1 second's audio data, so we need to calculate the first sample end time in processingBuffer
+							sampleEnd := processingBufferStart + sampleDuration
+
+							//Compare first sample end time with the end time of the specific response to determine when to stop removing
+							if sampleEnd > resultEndTime {
+								break
+							}
+
+							//Move the sample end time pointer
+							processingBufferStart = sampleEnd
+							//Remove the first sample(audio chunk) from the head of processingBuffer like Queue(FIFO)
+							processingBuffer = processingBuffer[1:]
+						}
+
+					}
+				}
+
+			}
+			//If there is no more data to read from input, then the program will terminate
+			if inputIsNull {
```

Comment:
Can `return inputIsNull` directly. No need for the `if`. But, I'm not sure I understand how `inputIsNull` works here. Might be better after it's split into more than one function.

Reply:
inputIsNull is a flag variable which will be changed by a concurrent function.  It is used to determine if the program needs to terminate.  We need "if" clause to return right decision whether to exit the for loop or not.

Reply:
```
return inputIsNull
```

is equivalent to:

```
if inputIsNull {
	return true
} else {
	return false
}
```